### ISSUE TITLE
Use Prismic's GraphQL API?

### ISSUE THREAD

**Issue by seaneking** [angeloashmore/gatsby-source-prismic#217]
Just wondering if there's any plans to refactor this source to use Prismic's official GraphQL API? I generally much prefer the approach this plugin is taking to [gatsby-source-prismic-graphql](https://github.com/birkir/gatsby-source-prismic-graphql), but not using the GraphQL API is a dealbreaker. It means that you have to provide your own schemas, rather than inferring them, and that previews are much harder to work with.

I understand that it would be a big rewrite though, so just wanted to see if it had been considered.

**Comment by angeloashmore** [comment]
Hi @seaneking, thanks for bringing this up. Moving to the GraphQL API has not been a strong consideration, but I am not against using it. Do you have ideas of how the GraphQL API could be utilized to improve the plugin?

These are some of the benefits of this plugin over using Prismic's GraphQL API directly:

- A focus on doing as much work as possible at build time (e.g. converting Rich Text to HTML).
- A more standard internal GraphQL API (i.e. Gatsby's GraphQL layer) in my opinion.
- A less invasive data API. Changes to Prismic's GraphQL API, for example, could directly affect all sites using it compared to using an "adapter" like this plugin.
- Embraces Gatsby's hierarchy and infrastructure. Creating pages, for example, should not be a feature of a source plugin. New Gatsby features "just work" as a result, like incremental builds.

Generally speaking, purpose-built source plugins tend to have a better developer experience than ones using gatsby-source-graphql. I think this is because the purpose-built plugins can model the data and API UI specifically for Gatsby and its methodologies. The REST API just happens to be the way we fetch all of the source data.

As you alluded to, three of the largest pain points with the plugin are:

1. Setting up previews. It works, but it would be dishonest to say it's straightforward to setup and intuitive.
2. Providing schemas manually. Having a statically known schema resolved a lot users' issues, but it can be a hassle to copy/paste the JSON. Automatically grabbing them would be nice. I asked Prismic about this and there is no official API to fetch schemas, although GraphQL schema introspection may be an option.
3. Less than ideal developer experience due to a lack of early and descriptive errors/warnings and guides.

If there are ways to address those three points (or any other pain points you've experienced) through the use of the GraphQL API, while adhering to the benefits listed above, I am definitely open to discussion.

**Comment by seaneking** [comment]
Thanks for the detailed response @angeloashmore! I’ll touch on each of your points:

> A focus on doing as much work as possible at build time (e.g. converting Rich Text to HTML)

Personally I’d prefer my source to be as light and transparent as possible. Gatsby is SSRd so ‘build time’ is a bit of a misnomer. Prismic provides an official Richtext component for React that can be used to render its rich text objects directly, so you don’t need to litter your code with `_dangerouslySetInnerHTML`. Switching to the other source I found this to be a better experience.

> A more standard internal GraphQL API (i.e. Gatsby's GraphQL layer) in my opinion.

Again transparency is king. While there are bits of Prismic’s GraphQL API design I don’t love, I’d much rather be just following their official API schema and docs than that of a community source

> A less invasive data API. Changes to Prismic's GraphQL API, for example, could directly affect all sites using it 

See above. I’d prefer to build to official API specs and trust that Prismic wouldn’t just introduce breaking changes than build to a 3rd party plugin’s specs

> Embraces Gatsby's hierarchy and infrastructure. Creating pages, for example, should not be a feature of a source plugin. New Gatsby features "just work" as a result, like incremental builds.

100% agree. Rewriting that layer was a massive mistake by the other source, and one of the main reasons I’m looking to move away from it. See above notes about being light and transparent.

But all of the above points really aren’t a big deal, and more matters of opinion. The biggest issues for me are the ones you outlined — previews, schema inference, and not using real GraphQL data.

If this source used Prismic’s GraphQL API we wouldn’t need to manually copy and paste content models on every change, which is prone to user error and cumbersome. We wouldn’t need to fetch every property used by internals in queries, we’d get better warnings etc, we’d get new API endpoints/features/bug fixes/etc as soon as Prismic released them without having to do anything, and generally I feel like the dev exp would be more robust, transparent, and maintainable going forward.

**Comment by angeloashmore** [comment]
> Personally I’d prefer my source to be as light and transparent as possible. Gatsby is SSRd so ‘build time’ is a bit of a misnomer. Prismic provides an official Richtext component for React that can be used to render its rich text objects directly, so you don’t need to litter your code with _dangerouslySetInnerHTML. Switching to the other source I found this to be a better experience.

The converted HTML will be part of the static files, but the `<RichText>` component will run again on hydration. This is extra main-thread work that can be offloaded at build time.

This also adds Prismic-specific code into your components and makes it less portable. For example, if you had a template that accepts HTML as a prop, you could provide it data from a Markdown file via `childMarkdownRemark` or data from a Prismic document via `richTextField.html`. If the template instead accepted a Prismic Rich Text object, it would be limited to just that data type.

In the sites I work on, we convert the HTML to React components using [`react-html-renderer`](https://github.com/angeloashmore/react-html-renderer). This introduces the issue above with extra main-thread work, but it is more universal (accepts HTML rather than a proprietary format) and is an optional enhancement, not a requirement of the plugin.

The raw Prismic data is provided on the `raw` field of Rich Text fields and can be passed directly to `<RichText>` if desired.

> Again transparency is king. While there are bits of Prismic’s GraphQL API design I don’t love, I’d much rather be just following their official API schema and docs than that of a community source

> See above. I’d prefer to build to official API specs and trust that Prismic wouldn’t just introduce breaking changes than build to a 3rd party plugin’s specs

I can agree with that. There are some troubles with using the official GraphQL API directly, such as `gatsby-image` support and compile-time HTML. I know we can define custom GraphQL resolvers that provide both, but 1) it diverges from the official API, and 2) augmenting GraphQL schemas is a complex process, especially when introducing previews. Much of Gatsby's power is in the Node environment, which we lose when providing previews in the browser.

Prismic is the only popular CMS that I know of that provides run-time previews with Gatsby. Combining static and dynamic data is not a trivial process. This is why Gatsby Cloud, from what I understand, effectively provides previews by running `gatsby develop` in a special environment rather than supporting extra JS runtime previews.

Sorry if I sound close-minded - let's keep this conversation going. I'm definitely open to your idea, but I think there are some concerns to address. There have been development hurdles in `gatsby-source-prismic-graphql` due to its complexity re: running GraphQL queries in the browser that I would prefer to avoid.

**Comment by seaneking** [comment]
Hey not at all, this is a really productive discussion, and I definitely agree that there are issues with `gatsby-source-prismic-graphql` to be wary of. I do still think that leveraging the GraphQL API would be a net win and solve some real hurdles for this source though.

You're right about the rehydration hit for `<RichText />`. For me the minor DX improvement is acceptable with the minor rehydration cost (I'm assuming it's minor, I haven't actually checked), but totally valid choice to offload it to compile time 👍 

> There are some troubles with using the official GraphQL API directly, such as gatsby-image support and compile-time HTML

Yep again agree. The other source adds an extra query node at build time for `gatsby-image` for transforming with sharp based on a regex test, which is less than ideal. I think you could make depending on Prismic's Imgix a hard requirement for gatsby-image compat if that solves it, but tbh I haven't worked with Imgix much before so I'm not sure what's involved. Re: compile-time HTML, that's a bit trickier, and I'd assume it would need a GraphQL resolver.

I also agree that trying to provide out-of-the-box preview support with the GraphQL API directly has too many edge cases. I've run into so many of them with `gatsby-source-prismic-graphql`. I think this source's approach of providing low-level helpers for devs to build their own preview is a better approach, though I also think that leveraging GraphQL on the clientside would make that process smoother.

Basically I' hoping that utilising Prismic's GraphQL API would allow this source to overcome some of it's big limitations (touched on in the first comment), without needing to compromise on design choices — like compile-time HTML, preview helpers, Imgix `gatsby-image` compat, etc.

**Comment by seaneking** [comment]
Just wanted to say, after migrating away from this source in v2, I just did a spike rewriting one of my smaller landing pages with it and it is _so nice_ after the bug and edge-case riddled mess that `gatsby-source-prismic-graphql` is increasingly becoming. Thank you @angeloashmore 👏 

I wish this source was better known, because the Prismic docs now point to `gatsby-source-prismic-graphql` as the 'official' Gatsby source. Again, I think using GraphQL to take care of schema and internal field inference would really help there. That's prob my biggest remaining gripe (I've given up on Prismic previews for now)

**Comment by codingwithchris** [comment]
I just migrated to this plugin after having ENDLESS issues with `gatsby-source-prismic-graphql` and everything works perfectly. The only thing that took a bit of time was rewriting field names and typescript defs to account for some structural differences between the 2 plugins.

My only real hangup with this is definitely the manual scheme mapping.

I'm not sure if this would be related or not, but I build an introspection script for a NextJS/Apollo project for Prismic that auto-generated GraphQL schemas.

https://github.com/prismicio/apollo-link-prismic/issues/10

It is a cleaned-up version of the script found in `gatsby-source-prismic-graphql`. I built this specifically to handle slices, but it could be easily modified to get introspection data for all content types.

**Comment by angeloashmore** [comment]
Thanks @seaneking and @codingwithchris, glad it's working well!

I've been thinking more about how to take advantage of Prismic's GraphQL API and came to the following conclusions. I'd be happy to get your thoughts.

**Using the GQL API directly is effectively re-implementing `gatsby-source-graphql`**

`gatsby-source-graphql` already allows using a GraphQL API directly and could be a viable solution if you want to use the native API. I believe it's possible to augment the remote schema by providing custom resolvers for things like `gatsby-image` on Image fields.

I can see value in creating a layer above `gatsby-source-graphql` specific for Prismic. I believe that is what `gatsby-source-prismic-graphql` is doing.

**Proxying queries to the GQL API means we lose Gatsby's node dependency system**

Features like incremental builds would not be supported. There may be the possibility of creating nodes and hooking up page dependencies in the background manually, but could get messy. This may also require the need to create pages within the plugin which I don't feel is appropriate.

**Deriving custom type schemas from the GQL API does not look to be possible**

The GQL API does not provide enough type information to derive a sound GraphQL type. For example, an Embed field is resolved as a `Json` field in the GQL API, the same as Rich Text and Image fields. The JSON returned for each of those fields is, of course, shaped very differently from each other.

Please correct me if I'm wrong re: introspecting the types. I would love to automate fetching the schemas. It can be frustrating working with multiple developers simultaneously while someone is editing the schema as you're developing.

**Prismic provides a kind of custom type schema in the API Browser**

I explored this over a year ago before implementing the current schema solution (copy/pasting JSON files). If I remember correctly, this endpoint used to include slices, but it appears to be absent now.

This is an undocumented endpoint that I came across while combing through the API. Unfortunately, it is not useful because it is not an official API (i.e. it could be removed/altered at any time) and it does not include the whole schema.

```js
// https://reponame.prismic.io/apibrowser/types?ref=master

{
  "my.settings.site_copyright": ["StructuredText"],
  "my.settings.uid": ["UID"],
  "my.page.meta_title": ["Text"],
  "my.page.title": ["StructuredText"],
  "my.navigation.title": ["StructuredText"],
  "my.navigation.uid": ["UID"],
  "my.page.uid": ["UID"],
  "my.settings.site_name": ["StructuredText"],
  "my.settings.site_description": ["StructuredText"],
  "my.settings.title": ["StructuredText"]
}
```

---

## Side note

Types is primarily where Prismic's GraphQL API falls short:

1. Types are too wide due to frequent use of the JSON type and limits the ability to work with the GraphQL schema as data.
2. Because of the JSON type, subfields cannot be picked individually, leading to larger-than-necessary payloads.

Querying for a single Embed field gives the following data (pseudo query/response):

```graphql
query {
  page {
    body {
      ... on PageBodyEmbed {
        primary {
          video_embed
        }
      }
    }
  }
}
```

```graphql
{
  "video_embed": {
    "width": 480,
    "height": 270,
    "embed_url": "https://youtu.be/xxxxx",
    "type": "video",
    "version": "1.0",
    "title": "Sample video title",
    "author_name": "Sample author name",
    "author_url": "https://www.youtube.com/channel/xxxxx",
    "provider_name": "YouTube",
    "provider_url": "https://www.youtube.com/",
    "cache_age": null,
    "thumbnail_url": "https://example.com/xxxxx",
    "thumbnail_width": 480,
    "thumbnail_height": 360,
    "html": "<iframe width=\"480\" height=\"270\" src=\"https://www.youtube.com/embed/xxxxx?feature=oembed\" frameborder=\"0\" allow=\"accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>"
  }
}
```

If I just wanted the `html` field, I would have a bunch of unused extra data. Minifiers like Terser [don't remove unused properties from objects](https://github.com/terser/terser/issues/577).

**Comment by seaneking** [comment]
All sounds completely reasonable. Not being able to easily derive schemas from the GQL API is a bummer, but really that's my only remaining gripe here. The reasons you've outlined against leveraging GQL for Prismic outweigh those benefits, so happy to close this one and move on.

FWIW though maybe we could add an explainer covering some of this (at a higher level) in the README? Something like

> #### Differences to `gatsby-source-prismic-graphql`
> {Basic outline of why it doesn't use GQL API and what that allows}

**Comment by seaneking** [comment]
Just stumbled back across this — I think we can close it, as it's been explored pretty thoroughly and the current approach seems to be the best. The only remaining thing I'd suggest is adding some documentation around these shortfalls and differences to the README, because I think a fair few people coming to `gatsby-source-prismic` will be refugees from `gatsby-source-prismic-graphql` like myself and @codingwithchris
